### ISSUE TITLE
Cannot access uninitialized non-nullable property when accessing a non-db field property of a document

### ISSUE THREAD

**Issue by andythorne** [doctrine/mongodb-odm#2349]
### Bug Report

|    Q        |   A
|------------ | ------
| BC Break    | yes
| Version     | 2.2.1

#### Summary
This happens on PHP >= 7.4 where the Document properties have been type hinted.

A fatal error is thrown when you access a property of a Proxy document where the property itself is NOT a field managed by the ODM, even if that property has a default value. This only seems to happen when the Proxy document has been loaded directly from the ODM (i.e. NOT from a __get call from accessing $document->someProperty).

#### Current behaviour
Throws a fatal error when property is accessed:
```
  [Error]                                                                                              
  Cannot access uninitialized non-nullable property App\Document\Customer::$domainEvents by reference  
```

#### How to reproduce
Reproduced on this repo: https://github.com/andythorne/doctrine-odm-example-error

Key bit copied from above repo. See it's [documents](https://github.com/andythorne/doctrine-odm-example-error/tree/main/src/Document) and [command to reproduce the error](https://github.com/andythorne/doctrine-odm-example-error/blob/main/src/Command/RunTestCommand.php)
```php
// The Order is loaded directly via the DM. The returned document is an Order object, as expected. The Order
// object has a customer property that contains a Proxy<Customer> object.
/** @var Order $order */
$order = $this->documentManager->getRepository(Order::class)->find($orderId);

// Fetch a list of Customers from the DB. Any customer object that was referenced in the above order is still
// a proxy object, however it will not have the defaults set for the un-managed $domainEvents property.
$customers = $this->documentManager->getRepository(Customer::class)->findAll();

/** @var Customer $customer */
foreach ($customers as $customer) {
    $customer->doSomeUpdate(); // This will trigger an error as we try to append to the non-existing $domainEvents property
}
```

#### Expected behaviour
Any properties should have their default values set from the Document class.

**Comment by webmozart** [comment]
I'm having the same problem, unfortunately.

**Comment by ozahorulia** [comment]
I'm having the same problem with PHP 8-style constructors:

```
    public function __construct(
        private ?array $userMetaData = null
    ) {
```

**Comment by Steveb-p** [comment]
@Hast constructors are not called by Doctrine initializer as far as I know, unless something changed recently.

Reason being this allows you to have your own constructors with enforced data, while Doctrine is "only" interested in setting the relevant fields with values coming from database.

If you need to do some set up when Doctrine loads an object, then you can use PostLoad lifecycle event.

**Comment by IonBazan** [comment]
Exactly, like @Steveb-p said - when Doctrine hydrates your entities via Proxy objects, constructor is not called because the remaining data should be hydrated from DB. This is expected behaviour and applies to ORM AFAIK. 

I understand that sometimes you may need to set some missing stuff in unmanaged properties to make your Document class "valid" but you may achieve that by using `PostLoad` callback (and calling it again from constructor when you create the instance manually).

```php
#[Document]
#[HasLifecycleCallbacks]
class Customer
{
    public function __construct()
    {
        $this->init();
    }

    #[PostLoad]
    public function init(): void
    {
        $this->myField = 'test';
    }
}
```

**Comment by ozahorulia** [comment]
@Steveb-p I understand that constructor is not called, but in this case those are not constructor logic or initialization, but a new way of declaring properties as of PHP 8. That's why I suppose It must follow the same logic as with old-style property declaration, mustn't it?

**UPD:**

Sorry, my initial code snippet was incorrect. What I meant was:
```
    public function __construct(
        private ?array $userMetaData = null
    ) {
 ```

**Comment by malarzm** [comment]
I'm afraid this needs to be solved on https://github.com/Ocramius/ProxyManager level

**Comment by alcaeus** [comment]
This should be fixed by #2698.
